lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -1928,10 +1928,13 @@ void IRGenDebugInfoImpl::emitVariableDeclaration(
     if (Indirection)
       Operands.push_back(llvm::dwarf::DW_OP_deref);
 
-    // There are variables without storage, such as "struct { func foo() {}
-    // }". Emit them as constant 0.
-    if (isa<llvm::UndefValue>(Piece))
-      Piece = llvm::ConstantInt::get(IGM.Int64Ty, 0);
+    // There are undefined values. Emit them as constant 0.
```

Comment:
The idea behind the old code was to trick LLVM into emitting variable definitions for zero-sized values. If the SIL undef value has a size > 0 then it would be better to emit an llvm::UndefValue here, too, since we don't want to give the impression that an uninitialized value is zero.

Reply:
For the record: Ideally, just anchoring the storage-less variable in the list of retainedEntities in the DISubprogram without an explicit dbg.value is *supposed* to be enough. I don't recall whether this was a workaround for LLVM or for LLDB (probably the latter).

Reply:
I'm not sure how that used to work, though, since AFAICT the code works by going through the explosion schema and emitting debug info for each exploded component of the aggregate. If a variable definition has a zero-sized value, then it won't have any explosion components at all, so this would never be reached.

Reply:
Note that the array of values this loop iterates over isn't necessarily an explosion. The emitVariableDeclaration function is called in several places. I would expect that the `test/DebugInfo/nostorage.swift` testcase also triggers this code.

In any case: I'd be happy if the nostorage.swift testcase keeps working, and a SILUndefValue with a non-zero storage size causes an llvm.dbg.value(undef, ...) to be emitted.